backends/candle/src/models/qwen3.rs

```diff
@@ -18,3 +22,577 @@ pub struct Qwen3Config {
+            let repeat_factor = self.num_attention_heads / self.num_key_value_heads;
+            let (b, h, s, d) = k.shape().dims4()?;
+            let k = k.unsqueeze(2)?.expand((b, h, repeat_factor, s, d))?;
+            k.reshape((b, h * repeat_factor, s, d))?
+        } else {
+            k
+        };
+        let v = if self.num_key_value_heads != self.num_attention_heads {
+            let repeat_factor = self.num_attention_heads / self.num_key_value_heads;
+            let (b, h, s, d) = v.shape().dims4()?;
+            let v = v.unsqueeze(2)?.expand((b, h, repeat_factor, s, d))?;
+            v.reshape((b, h * repeat_factor, s, d))?
+        } else {
+            v
+        };
+
+        let attention_scores = q.matmul(&k.t()?)?;
+        let mut attention_scores = (attention_scores * self.softmax_scale)?;
+
+        if let Some(attention_bias) = attention_bias {
+            attention_scores = attention_scores.add(attention_bias)?;
+        }
+
+        let attention_probs = candle_nn::ops::softmax_last_dim(&attention_scores)?;
+        let context_layer = attention_probs.matmul(&v.contiguous()?)?;
+
+        // Transpose back and flatten: [batch, heads, seq_len, head_dim] -> [batch, seq_len, hidden_size]
+        let context_layer = context_layer.transpose(1, 2)?.flatten_from(D::Minus2)?;
+
+        self.o_proj.forward(&context_layer)
+    }
+}
+
+struct Qwen3MLP {
+    gate_up_proj: Linear,
+    down_proj: Linear,
+
+    intermediate_size: usize,
+
+    span: tracing::Span,
+}
+
+impl Qwen3MLP {
+    pub fn load(vb: VarBuilder, config: &Qwen3Config) -> Result<Self> {
+        let intermediate_size = config.intermediate_size;
+
+        let gate_proj_weight = vb
+            .pp("gate_proj")
+            .get((intermediate_size, config.hidden_size), "weight")?;
+
+        let up_proj_weight = vb
+            .pp("up_proj")
+            .get((intermediate_size, config.hidden_size), "weight")?;
+
+        let gate_up_proj_weight = Tensor::cat(&[&gate_proj_weight, &up_proj_weight], 0)?;
+        let gate_up_proj = Linear::new(gate_up_proj_weight, None, None);
+
+        let down_proj_weight = vb
+            .pp("down_proj")
+            .get((config.hidden_size, intermediate_size), "weight")?;
+        let down_proj = Linear::new(down_proj_weight, None, None);
+
+        Ok(Self {
+            gate_up_proj,
+            down_proj,
+            intermediate_size,
+            span: tracing::span!(tracing::Level::TRACE, "mlp"),
+        })
+    }
+
+    pub fn forward(&self, hidden_states: &Tensor) -> Result<Tensor> {
+        let _enter = self.span.enter();
+
+        let gate_up_states = self.gate_up_proj.forward(hidden_states)?;
+        let gate_states = gate_up_states.narrow(D::Minus1, 0, self.intermediate_size)?;
+        let up_states = gate_up_states.narrow(D::Minus1, self.intermediate_size, self.intermediate_size)?;
+
+        let gate_states = gate_states.silu()?;
```

Comment:
Could you leverage the `HiddenAct.forward` method instead? i.e.

```suggestion
        let gate_states = self.act.forward(&gate_states)?;
```